pkg/controller/master/upgrade/upgrade_controller.go

```
@@ -130,6 +132,23 @@ func (h *upgradeHandler) OnChanged(_ string, upgrade *harvesterv1.Upgrade) (*har
 
 	repo := NewUpgradeRepo(h.ctx, upgrade, h)
 
+	isRestoreVM, err := util.IsRestoreVM()
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"namespace": upgrade.Namespace,
+			"name":      upgrade.Name,
+		}).WithError(err).Error("Failed to get setting UpgradeConfig")
+		return nil, err
```

Will this cause upgrade to abort or stuck in a loop if `err` doesn't goes away?

---

Yes, it will stuck forever. Perhaps we can ignore the error here as we already have a validator for upgradeconfig to make sure the json format is valid. https://github.com/harvester/harvester/blob/master/pkg/webhook/resources/setting/validator.go#L1620-L1630

---

the `restoreVM` is out of the upgrade controller `OnChange` main state machine,  when error returns, it shortcuts the state machine, causes the latter can't move forward; if a better place can't be found, then log error, enqueue and continue following code will be the solution. WDYT ?

---

> the restoreVM is out of the upgrade controller OnChange main state machine, when error returns, it shortcuts the state machine, causes the latter can't move forward; if a better place can't be found, then log error, enqueue and continue following code will be the solution. WDYT ?

I assume you mean add `h.upgradeController.EnqueueAfter(upgrade.Namespace, upgrade.Name, time.Second*10)` if util.IsRestoreVM raise error, right?

> Just noticed the upgrade.Labels[harvesterLatestUpgradeLabel] == "true" is checked way after; but for restoreVM, it strictly depends on this label, to avoid confused by legacy upgrade objects

Good catch! Thanks!

@w13915984028, I've addressed all comments in the latest commit, hopefully I didn't get something wrong.